### ISSUE TITLE
User Selected Payout Network

### ISSUE THREAD

**Issue by pavlovcik** [ubiquity/ubiquibot#589]
It has come to my attention that some contributors are confused about cashing out from Gnosis Chain. I've already received three notifications from three different contributors for context. 

We should default to Mainnet payouts and then allow users to opt in to Gnosis Chain payouts to save on gas. 

This issue needs sub tasks to roll out support for this but I am unsure what exactly needs to be done.

**Comment by pavlovcik** [comment]
@rndquu do you think you can help me spec this out?

**Comment by rndquu** [comment]
> @rndquu do you think you can help me spec this out?

We could add a new command similar to `/payout-network NETWORK_ID` which would save the desired network to a DB.

So if bounty hunter has an empty "payout network" in a DB then use network from the bot's config else use the one provided by a bounty hunter.

What do you think?

**Comment by pavlovcik** [comment]
We can consider Gnosis style addresses upon wallet registration e.g. `eth:0x4007CE2083c7F3E18097aeB3A39bb8eC149a341d`, `op:0x4007CE2083c7F3E18097aeB3A39bb8eC149a341d`, `gno:0x4007CE2083c7F3E18097aeB3A39bb8eC149a341d`

**Comment by rndquu** [comment]
> I think we should always default to Mainnet because new people definitely know how to use at least Mainnet before they learn how to use sidechains/layer 2s.
> 
> We can consider [EIP3770](https://eips.ethereum.org/EIPS/eip-3770) style addresses upon wallet registration e.g:
> 
> * `eth:0x4007CE2083c7F3E18097aeB3A39bb8eC149a341d`
> * `op:0x4007CE2083c7F3E18097aeB3A39bb8eC149a341d`
> * `gno:0x4007CE2083c7F3E18097aeB3A39bb8eC149a341d`
> 
> I have some concerns though:
> 
> * It would be the best developer experience if we enforced that all of our partners top up funds on both Mainnet and Gnosis Chain (annoying for partners)
> * If a bounty hunter has their wallet registered in our network using Gnosis Chain, and if a new partner does not have any funds on Gnosis Chain, that can get really annoying for everybody.
> 
> Not sure if it makes sense to offer a service to handle this setup for our partners? For example, if they don't want to deal with setting up their own wallet, we could handle it for partners as a convenience and charge an extra fee? I could see lazy partners finding that appealing. We just need to let them know to top up regularly, and that we do not assume any responsibilities for hacked/lost funds.

Right now partners support payouts only for a single [network](https://github.com/ubiquity/ubiquibot/blob/development/.github/ubiquibot-config.yml#L1C1-L1C15).

We could modify the bot's `evm-network-id` param to accept an array of supported networks for payouts. Example:
```
evm-network-ids: [1, 100]
```

This way when a bounty hunter tries to set a specific network for payouts (via `/payout-network NETWORK_ID`) then:
 - If the specified by a bounty hunter network is set in the partner's config then allow to set the network else show an error `Network is not supported by a partner's project`

**Comment by pavlovcik** [comment]
> We could modify the bot's evm-network-id param to accept an array of supported networks for payouts. 

I think that's a great idea. Let's do it!

**Comment by rndquu** [comment]
```
We could modify the bot's evm-network-id param to accept an array of supported networks for payouts. 

Example:
evm-network-ids: [1, 100]
```

I've checked the bot's code and this feature requires quite a refactor (penalties, permits). But what's worse it adds ambiguity (in the code) because it isn't really clear what network is used at the moment if we transform `evm-network-id` to array.

We could make it simple:
1. Add a new "bounty hunter specifies network for payouts" command (i.e. `/payout-network NETWORK_ID`)
2. Disable the `/payout-network` command in the [default config](https://github.com/ubiquity/ubiquibot/blob/development/ubiquibot-config-default.json) (i.e. for all our partners)
3. Enable the `/payout-network` command in the ubiquity org wide [config](https://github.com/ubiquity/ubiquibot-config/blob/development/.github/ubiquibot-config.yml)

This way when a new partner adds the bot to his project then he would:
1. Specify `evm-network-id` for payouts
2. If partner wants to support payouts for multiple networks then he would need to enable the `/payout-network` command and [setup](https://dao.ubq.fi/ubiquibot-setup) the wallet accordingly

In the context of the ubiquity org:
- we have both payout network setups ready (mainnet + gnosis)
- when a bounty hunter calls `/payout-network NETWORK_ID` then this specified network id is saved to a DB, so the payout network precedence is this one (from the least important to the most important):
  1) Bot's default config
  2) Org wide config
  3) Repo wide config
  4) Network from DB which bounty hunter specified via `/payout-network NETWORK_ID`

**Comment by pavlovcik** [comment]
I'm still thinking on your proposal but you did just inspire me to come up with something new that is somewhat relevant to this. 

What if we only generate payment permits when bounty hunters want to cash out? 

For example:

1. A bounty hunter solves issue 1. 2, and 3. 
2. A bot comment is appended to each of them saying something like "you earned 50 DAI, you can claim here (link)"
3. The total owed amount is saved in our database
4. When the bounty hunter goes to pay.ubq.fi, the amount we owe them is totaled up and displayed example: "Claim 150 DAI"
5. The claim button generates a payment permit of 150 DAI for them to use. 

This flow allows them to select pay out network conveniently. 

It also rolls up payments into a single transaction.

**Comment by rndquu** [comment]
> 5. The claim button generates a payment permit of 150 DAI for them to use.

This requires a backend server because we somehow need to show a signed permit to a bounty hunter.

We could (keeping in mind that we save the owed amount in a DB) also introduce a new command `/payout` (as far as I remember we already have this command but it simply regenerated a permit for an issue) which:
1. Generates a permit URL
2. Resets owed amount in a DB

This way we don't even need to modify https://pay.ubq.fi/ (but still need to modify https://pay.ubq.fi/audit)

**Comment by pavlovcik** [comment]
> > 5. The claim button generates a payment permit of 150 DAI for them to use.
> 
> 
> 
> This requires a backend server because we somehow need to show a signed permit to a bounty hunter.
> 

Cloudflare Worker? I want to clarify that this all happens inside of the dApp

> 
> We could (keeping in mind that we save the owed amount in a DB) also introduce a new command `/payout` (as far as I remember we already have this command but it simply regenerated a permit for an issue) which:
> 
> 1. Generates a permit URL
> 
> 2. Resets owed amount in a DB
> 
> 
> 
> This way we don't even need to modify https://pay.ubq.fi/ (but still need to modify https://pay.ubq.fi/audit)

**Comment by rndquu** [comment]
> Cloudflare Worker? I want to clarify that this all happens inside of the dApp.

Cloudflare worker is basically a backend server

> I think it's more elegant for bounty hunters to manage their finances entirely inside of that dApp, instead of relying on both the GitHub comment UI and then the dApp.

This way bounty hunters still rely on 2 parties: the dApp and cloudflare worker. What other benefits do we get (besides "managing finances in a single place") moving from generating permits on the bot's side to generating permits in a cloudflare worker? I can not make one. But we do get a new service to maintain (i.e. more work to do). 

Again, I don't understand why we should break the feature that is working fine and everybody seems to be happy with it. It would be more clear (in terms of what other cases should "payout worker" used for) if we had a "payouts roadmap", like:
1. Q1 2023: cloudflare worker to generate permits
2. Q2 2023: bounty hunter dashboard at https://pay.ubq.fi/
3. Q3 2023: issues matchmaking
4. Q4 2024: partner dashboard at https://pay.ubq.fi/

Right now a separate "permit cloudflare worker" looks like an unnecessary refactor

**Comment by pavlovcik** [comment]
> But we do get a new service to maintain (i.e. more work to do).

We can remove the original flow so that we don't need to maintain two methods of payout. 

> Again, I don't understand why we should break the feature that is working fine and everybody seems to be happy with it.

Breaking changes sometimes are necessary to evolve. That's perhaps why semver refers to the first digit as "breaking changes".  

Utility is certainly important to achieve, but once we get closer to "version one" territory and beyond, a lot of our work will be about small improvements to enhance the user experience, even if that means to refactor old features.  

> It would be more clear (in terms of what other cases should "payout worker" be used for) if we had a "payouts roadmap", like:

I'll have to think more on these permits related features specifically, but in regards to payments roadmap: we are actively looking to establish traditional banking infrastructure (fiat bank accounts, debit cards) in order for DAO contributors to be able to cash out more conveniently and bank with us. The stablecoin will of course be our on-chain backbone for all of this. 

It's important to keep in mind that in the future, we should be able to enable non-crypto natives to also be able to "work in the metaverse" and get paid. Ideally non-programmers as well! That's how we can make a billion dollar company. If the user experience is too crypto developer oriented, our offering will always remain extremely niche, which is not good for growth.

**Comment by rndquu** [comment]
@pavlovcik 

To sum up:
1. We save permits to a DB
2. We move permit generation logic to a separate cloudflare worker (i.e. https://pay.ubq.fi backend server)
3. We create a new UI at https://pay.ubq.fi/ for rolled up payouts

Right?

**Comment by pavlovcik** [comment]
1. Just save the balance on database and then generate the single permit for the entire owed amount at payment time.

**Comment by pavlovcik** [comment]
> @pavlovcik 
> 
> 
> 
> To sum up:
> 
> 1. We save permits to a DB
> 

Only the owed amount. For auditing purposes it might be wise to also associate the original issue there. 

> 2. We move permit generation logic to a separate cloudflare worker (i.e. https://pay.ubq.fi backend server)
> 

Yes

> 3. We create a new UI at https://pay.ubq.fi/ for rolled up payouts
> 

I think we just need to adjust the main page to 
1. read the owed balance from the database 
2. Generate a payment permit for them to claim when they press the claim button. My only concern here is if the claim transaction fails for any reason and they leave the page, will they forever lose that payment permit?

**Comment by rndquu** [comment]
> My only concern here is if the claim transaction fails for any reason and they leave the page, will they forever lose that payment permit?

> I'm concerned now about them generating permits for multiple networks. We will need to think carefully about the permit generation flow here

We'll find out the solution

Ok, I will break it down into subtasks

**Comment by pavlovcik** [comment]
> 2. Generate a payment permit for them to claim when they press the claim button. My only concern here is if the claim transaction fails for any reason and they leave the page, will they forever lose that payment permit?

This is a naive proposal but perhaps there is a way to generate a nonce by hashing the queue of owed transactions. I'm not entirely sure how this is possible but this is somewhat inspired by my understanding of merkle tries, and how they hash hashes to ensure data integrity. Each transaction in the queue can be hashed when a permit is generated for it, and then the queue of transaction hashes can be hashed as well. Perhaps in this way there is some way to verify if a specific payout (for a specific issue) was rolled up inside of a permit, so that double spending (double permit issuance for the same issue) is  
not possible?

**Comment by pavlovcik** [comment]
1. That's unfortunate. I would like to have a think on this longer to see if we can minimize the amount of clicks required to cash out on that UI for a superior user experience. 
2. The last option seems to be the best because it relies mostly on GitHub's security infrastructure.

**Comment by rndquu** [comment]
> Is it possible that we try every partner and only the valid keys make it through?

I think it could work (though this is a weird approach) but what if a bounty hunter has unclaimed permits in multiple organizations? Ideally bounty hunter should claim payouts from multiple organizations in a single transaction which brings us back to a [custom](https://github.com/ubiquity/pay.ubq.fi/issues/101) permit2 contract with multicall support. Anyway I think it's ok to select an organization at https://pay.ubq.fi.

> Lastly I think we would benefit from a micro service (we can set up a Cloudflare Worker with an API) that checks if permits are claimed.

We could simply add it as a new API call for the "pay.ubq.fi permit generation worker"

**Comment by pavlovcik** [comment]
> Anyway I think it's ok to select an organization at https://pay.ubq.fi/.

It would be a good user experience if there is a radio selection from a list of organizations which owe payments. For example

[ ubiquity ] [ alpha labs ] 

And the user selects one at a time. Also the radio option disappears once no more owed payments exist.

**Comment by pavlovcik** [comment]
Draft schema for 

>  I'm starting to think that we would benefit from keeping track of every permit

```json
{
    "payments": [
        {
            "id": "98d262a7368632785573f1bcdc153bc3",
            "organization": "ubiquity",
            "repository": "ubiquibot",
            "issue": "100",
            "amount": "100000000000000000000",
            "type": "assignee"
        },
        {
            "id": "a3da2829fe749f25a97cdb1e1cfef991",
            "organization": "ubiquity",
            "repository": "ubiquibot",
            "issue": "104",
            "amount": "10000000000000000000",
            "type": "comments"
        },
        {
            "id": "82add8958de8fa4482d7177b0a5b3877",
            "organization": "ubiquity",
            "repository": "ubiquibot",
            "issue": "105",
            "amount": "10000000000000000000",
            "type": "creator"
        },
        {
            "id": "934badf734685398db9c7cfbd22cda4b",
            "organization": "ubiquity",
            "repository": "ubiquibot",
            "issue": "105",
            "amount": "5000000000000000000",
            "type": "comments"
        },
        {
            "id": "82add8958de8fa4482d7177b0a5b3877",
            "organization": "ubiquity",
            "repository": "ubiquibot",
            "issue": "106",
            "amount": "50000000000000000000",
            "type": "reviewer"
        }
    ],
    "permit": {
        "permitted": {
            "token": "0xe91D153E0b41518A2Ce8Dd3D7944Fa863463a97d",
            "amount": "175000000000000000000"
        },
        "nonce": "52856288763743795411125691792278824287979693722545410590222942625942325296142",
        "deadline": "115792089237316195423570985008687907853269984665640564039457584007913129639935"
    },
    "transferDetails": {
        "to": "0x4007CE2083c7F3E18097aeB3A39bb8eC149a341d",
        "requestedAmount": "175000000000000000000"
    },
    "owner": "0xf87ca4583C792212e52720d127E7E0A38B818aD1",
    "signature": "0x0d89e8dfa2035eff7b4091277c9fb5f240fd25d48e67cacee22eb9d7986c4277192310fda562e558608481dcc13d4ec75026483343182adcb1e77b9d65df477b1b"
}
```

**Comment by rndquu** [comment]
@pavlovcik 

There are 2 main contracts in uniswap's permit2 repo: [SignatureTransfer](https://docs.uniswap.org/contracts/permit2/reference/signature-transfer) and [AllowanceTransfer](https://docs.uniswap.org/contracts/permit2/reference/allowance-transfer)

They have subtle differences:
- `SignatureTransfer`: transfers tokens on permit claim and has **unordered** nonces
- `AllowanceTransfer`: sets permitted allowance on permit claim and has **ordered** nonces per `[owner][token][spender]`

Right now we use `SignatureTransfer` for the permit generation mainly because of the unordered nonces feature

We could move from `SignatureTransfer` to `AllowanceTransfer`. This way we don't need to think of the best way to generate nonces because we can simply take them on-chain.

Example using the `AllowanceTransfer` contract:
1. Bounty hunter solves 2 issues (worth 50 USD and 100 USD)
2. We save in a DB that we owe that bounty hunter 150 USD
3. Bounty hunter opens https://pay.ubq.fi and clicks on a `generate permit` button
4. Cloudflare worker takes the current bounty hunter's nonce **on-chain** and generates a permit that sets a bounty hunter's allowance to 150 USD
5. Bounty hunter leaves https://pay.ubq.fi without redeeming the permit
6. Bounty hunters solves the 3rd issue worth 200 USD, we owe him 350 USD now
7. Bounty hunter again opens https://pay.ubq.fi and clicks on a `generate permit` button
8. Cloudflare worker takes the current bounty hunter's nonce (which hasn't changed since step 4) on-chain and generates a permit that sets a bounty hunter's allowance to 350 USD

In this scenario bounty hunter has 2 valid permits but:
1. He can redeem only 1 of them because of the nonce
2. It is more profitable for him to redeem the latest

**Comment by pavlovcik** [comment]
> Right now we use SignatureTransfer for the permit generation mainly because of the unordered nonces feature

What is the trade off if we switch to ordered nonces? I don't understand the benefits of unordered nonces in our context.

**Comment by rndquu** [comment]
> > Right now we use SignatureTransfer for the permit generation mainly because of the unordered nonces feature
> 
> What is the trade off if we switch to ordered nonces? I don't understand the benefits of unordered nonces in our context.

Thinking more of it both `SignatureTransfer` and `AllowanceTransfer` are fine in our case because in the end we need a DB for both of the approaches so we'd better use `SignatureTransfer` because we already have a codebase for it.

P.S. I wanted to find out an on-chain solution but since we save permits in a DB at least for auditing purposes it doesn't make much sense to stick to `AllowanceTransfer`

**Comment by pavlovcik** [comment]
> P.S. I wanted to find out an on-chain solution but since we save permits in a DB at least for auditing purposes it doesn't make much sense to stick to `AllowanceTransfer`

Can you elaborate? I think audibility is very important.

**Comment by rndquu** [comment]
> > P.S. I wanted to find out an on-chain solution but since we save permits in a DB at least for auditing purposes it doesn't make much sense to stick to `AllowanceTransfer`
> 
> Can you elaborate? I think audibility is very important.

I meant that since we save permits in a DB (for auditing purposes) then it doesn't make much sense trying to find a fully on-chain solution (i.e. it doesn't matter what we use `SignatureTransfer` or `AllowanceTransfer`) because in the end we still need a DB for this case:
1. Bounty hunter generates a permit and closes pay.ubq.fi without redeeming
2. Bounty hunter solves another issue
3. Bounty hunter generates a permit again

On step 3 we need to modify the permit generated on step 1 (because we need to add the reward from step 3 to the reward from step 1). If we stick to the `SignatureTransfer` (this is the plan) then we use the permit's DB row `id` from step 1 as a nonce. If we stick to the `AllowanceTransfer` (and use an ordered on-chain nonce that this contract provides) then we still need to save an on-chain nonce from step 1 to a DB because we need to check whether it'd been earlier redeemed to find out whether we should regenerate the last permit or create a new one. So in the end both `SignatureTransfer` and `AllowanceTransfer` require a DB.

TLDR;
Again, we can use both `SignatureTransfer` and `AllowanceTransfer` but `SignatureTransfer` is better because we have an existing codebase for it.

**Comment by pavlovcik** [comment]
> TLDR;
> Again, we can use both SignatureTransfer and AllowanceTransfer but SignatureTransfer is better because we have an existing codebase for it.

I disagree because based on your [comment](https://github.com/ubiquity/ubiquibot/issues/589#issuecomment-1684897224), you state that ordered nonces will allow us to "roll up" permits which is a superior solution. 

I think it's fine to throw out obsolete code if it detracts from improving the product.

**Comment by rndquu** [comment]
> I disagree because based on your https://github.com/ubiquity/ubiquibot/issues/589#issuecomment-1684897224, you state that ordered nonces will allow us to "roll up" permits which is a superior solution.

Both `SignatureTransfer` and `AllowanceTransfer` will allow us to "roll up" permits because the total owed amount is stored in a DB.

**Comment by pavlovcik** [comment]
I was beginning to write a long question but while writing it out and thinking through it, it all makes sense to me now. Let's do it.

**Comment by rndquu** [comment]
> Would you mind creating a new issue specification describing this?

Sure

> If they try and regenerate on the same network, then we load the old permit worth some money

Yes

> If they try and generate on the new network, the permit generation fails because the owed amount is 0

Yes, it makes sense to throw an error and refuse to generate a permit for a different network if the previous permit (for another chain) has not yet been claimed.

**Comment by rndquu** [comment]
@pavlovcik 

The bot itself is a server app so instead of deploying a separate cloudflare worker we could also add the needed API methods directly to the bot (i.e. add a few API methods to the bot so that it could respond to github events along with API calls from https://pay.ubq.fi). 

This way we still save the total owed amount in a DB and have a single UI for rewards redeeming (i.e. https://pay.ubq.fi) but drastically save time on moving the permit generation logic to a separate service because the bot has already implemented everything.

**Comment by pavlovcik** [comment]
That sounds like a plan. Be sure to include that in your new specification. I assume that means we can close this current issue as not completed

**Comment by rndquu** [comment]
> That sounds like a plan. Be sure to include that in your new specification. I assume that means we can edit the original comment of this issue to be a checklist of all of the tasks required to make this happen now. Are you able to do that?

Sure

> We might have to close out some other issues. I noticed when I was on https://github.com/ubiquity/ubiquibot/issues?q=gnosis+safe that we have some issues related to other network payouts, like https://github.com/ubiquity/ubiquibot/issues/347

I'll check it
